x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/slm/SnapshotRetentionTask.java

```diff
@@ -290,117 +265,81 @@ static String getPolicyId(SnapshotInfo snapshotInfo) {
-                final long deleteStartTime = nowNanoSupplier.getAsLong();
-                // TODO: Use snapshot multi-delete instead of this loop if all nodes in the cluster support it
-                //       i.e are newer or equal to SnapshotsService#MULTI_DELETE_VERSION
-                deleteSnapshot(policyId, repo, info.snapshotId(), slmStats, ActionListener.wrap(acknowledgedResponse -> {
-                    deleted.incrementAndGet();
-                    assert acknowledgedResponse.isAcknowledged();
-                    historyStore.putAsync(SnapshotHistoryItem.deletionSuccessRecord(Instant.now().toEpochMilli(),
-                            info.snapshotId().getName(), policyId, repo));
-                }, e -> {
-                    failed.incrementAndGet();
-                    try {
-                        final SnapshotHistoryItem result = SnapshotHistoryItem.deletionFailureRecord(Instant.now().toEpochMilli(),
-                            info.snapshotId().getName(), policyId, repo, e);
-                        historyStore.putAsync(result);
-                    } catch (IOException ex) {
-                        // This shouldn't happen unless there's an issue with serializing the original exception
-                        logger.error(new ParameterizedMessage(
-                            "failed to record snapshot deletion failure for snapshot lifecycle policy [{}]",
-                            policyId), ex);
-                    }
-                }));
-                // Check whether we have exceeded the maximum time allowed to spend deleting
-                // snapshots, if we have, short-circuit the rest of the deletions
-                long finishTime = nowNanoSupplier.getAsLong();
-                TimeValue deletionTime = TimeValue.timeValueNanos(finishTime - deleteStartTime);
-                logger.debug("elapsed time for deletion of [{}] snapshot: {}", info.snapshotId(), deletionTime);
-                TimeValue totalDeletionTime = TimeValue.timeValueNanos(finishTime - startTime);
-                if (totalDeletionTime.compareTo(maximumTime) > 0) {
-                    logger.info("maximum snapshot retention deletion time reached, time spent: [{}]," +
-                            " maximum allowed time: [{}], deleted [{}] out of [{}] snapshots scheduled for deletion, failed to delete [{}]",
-                        totalDeletionTime, maximumTime, deleted, count, failed);
-                    slmStats.deletionTime(totalDeletionTime);
-                    slmStats.retentionTimedOut();
-                    return;
-                }
+            deleteSnapshots(slmStats, deleted, failed, repo, snapshots, allDeletesListener);
+        }
+    }
+
+    /**
+     * Set of all currently deleting {@link SnapshotId} used to prevent starting multiple deletes for the same snapshot.
+     */
+    private final Set<SnapshotId> runningDeletions = Collections.synchronizedSet(new HashSet<>());
```

Comment:
Can you move this up to the top before all the methods? It's a little strange to have a private class var randomly in the middle of the class

Reply:
Sure, force of habit. The snapshot classes in core eventually got so complex that this style started being more readable :D but here it's unnecessary :)